ui/packages/catalog-ui-search/src/main/webapp/js/widgets/cesium.base.line.js

```diff
@@ -0,0 +1,182 @@
+  constructor(options) {
+    super(...options);
+
+    Object.assign(this, {...options});
+  }
+
+  updatePrimitive = () => {
+    this.drawGeometry(this.model);
+  }
+
+  drawGeometry = (model) => {
+    throw new Error('This method must be overwritten by implementing geometry!');
+  }
+
+  constructLinePrimitive = (coordinates) => {
+    const color = this.model.get('color');
+
+    return {
+      width: 8,
+      material: Cesium.Material.fromType('PolylineOutline', {
+          color: color ? Cesium.Color.fromCssColorString(color) : Cesium.Color.KHAKI,
+          outlineColor: Cesium.Color.WHITE,
+          outlineWidth: 4
+      }),
+      id: 'userDrawing',
+      positions: Cesium.Cartesian3.fromDegreesArray(_.flatten(coordinates))
+    };
+  }
+
+  listenForCameraChange = () => {
+    this.map.scene.camera.moveStart.addEventListener(this.handleCameraMoveStart, this);
+    this.map.scene.camera.moveEnd.addEventListener(this.handleCameraMoveEnd, this);
+  }
+
+  handleCameraMoveStart = () => {
+    this.animationFrameId = window.requestAnimationFrame(function() {
+        if (needsRedraw(this)){
+            this.updatePrimitive();
+        }
+        this.handleCameraMoveStart();
+    }.bind(this));
+  }
+
+  handleCameraMoveEnd = () => {
+    window.cancelAnimationFrame(this.animationFrameId);
+    if (needsRedraw(this)) {
+        this.updatePrimitive();
+    }
+  }
+
+  destroy = () => {
+    if (!this.map || !this.map.scene) {
+      return;
+    }
+
+    this.map.scene.camera.moveStart.removeEventListener(this.handleCameraMoveStart, this);
+    this.map.scene.camera.moveEnd.removeEventListener(this.handleCameraMoveEnd, this);
+    window.cancelAnimationFrame(this.animationFrameId);
+    if (this.primitive) {
+        this.map.scene.primitives.remove(this.primitive);
+    }
+    this.remove(); // backbone cleanup.
+  }
+}
+
+class GeometryController extends DrawingController {
+  constructor(options) {
+    DrawingController.prototype.drawingType = options.drawingType;
+    DrawingController.prototype.options = options;
+    
+    super(...options);
+
+    Object.assign(this, {...options});
+  }
+
+  show = (model) => {
+    if (!this.enabled) {
+      return;
+    }
+    this.drawHelper.stopDrawing();
+    // remove old polygon
+    const existingView = this.getViewForModel(model);
+    if (existingView) {
+        existingView.destroy();
+        this.removeViewForModel(model);
+    }
+
+    this.addView(this.createRenderView(model));
+  }
+
+  createRenderView = (model) => {
+    throw new Error('This method must be implemented by the subclass!');
+  }
+
+  draw = (model) => {
+    const controller = this;
+    const toDeg = Cesium.Math.toDegrees;
+    if (!this.enabled) {
+        return;
+    }
+    // start geometry draw.
+    this.notificationView = new NotificationView({
+        el: this.notificationEl
+    }).render();
+
+    this.drawHelper[`startDrawing${capitalize(this.geometry)}`]({
```

Comment:
❓ Should we perform a check here to make sure that `this.geometry != undefined` because otherwise your anonymous function `capitalize` will break if you call `capitalize(undefined)`?

Reply:
@bellcc I agree, but in this scenario if we add the check and the `geometry` is undefined then nothing is drawn on the screen which would make it a pain to debug. I think you are probably meaning that we should handle the case of bad geometry with some sort of logging/feedback. @andrewkfiedler is there an application-wide error handler I can pass errors to?